services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,414 @@
+
+  if (checked) {
+    return available;
+  }
+
+  dbus::Bus::Options options;
+  options.bus_type = dbus::Bus::SYSTEM;
+  options.connection_type = dbus::Bus::PRIVATE;
+  auto bus = base::MakeRefCounted<dbus::Bus>(options);
+
+  dbus::ObjectProxy* proxy =
+      bus->GetObjectProxy(kServiceName, dbus::ObjectPath(kManagerObjectPath));
+
+  dbus::MethodCall call(kManagerInterfaceName, "GetClient");
+  auto response =
+      proxy->CallMethodAndBlock(&call, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT);
+
+  // If the response is |nullptr| then the GeoClue2.Manager does not exist.
+  available = response.get();
+  checked = true;
+
+  // Shutdown this bus - we'll use one on the DBus thread for our actual
+  // provider.
+  bus->ShutdownAndBlock();
+
+  return available;
+}
+
+}  // namespace
+
+struct GeoClueProperties : public dbus::PropertySet {
+  dbus::Property<std::string> desktop_id;
+  dbus::Property<uint32_t> requested_accuracy_level;
+
+  explicit GeoClueProperties(dbus::ObjectProxy* proxy)
+      : dbus::PropertySet(proxy, kClientInterfaceName, base::NullCallback()) {
+    RegisterProperty("DesktopId", &desktop_id);
+    RegisterProperty("RequestedAccuracyLevel", &requested_accuracy_level);
+  }
+
+  ~GeoClueProperties() override = default;
+};
+
+struct GeoClueLocationProperties : public dbus::PropertySet {
+  dbus::Property<double> latitude;
+  dbus::Property<double> longitude;
+  dbus::Property<double> accuracy;
+  dbus::Property<double> altitude;
+  dbus::Property<double> speed;
+  dbus::Property<double> heading;
+
+  explicit GeoClueLocationProperties(dbus::ObjectProxy* proxy)
+      : dbus::PropertySet(proxy, kLocationInterfaceName, base::NullCallback()) {
+    RegisterProperty("Latitude", &latitude);
+    RegisterProperty("Longitude", &longitude);
+    RegisterProperty("Accuracy", &accuracy);
+    RegisterProperty("Altitude", &altitude);
+    RegisterProperty("Speed", &speed);
+    RegisterProperty("Heading", &heading);
+  }
+
+  ~GeoClueLocationProperties() override = default;
+
+  void GetAllWithCallback(base::OnceCallback<void()> on_got_all) {
+    // We only support this one at a time. It's fine for now.
+    DCHECK(!on_got_all_);
+    on_got_all_ = std::move(on_got_all);
+    dbus::PropertySet::GetAll();
+  }
+
+  // dbus::PropertySet:
+  void OnGetAll(dbus::Response* response) override {
+    dbus::PropertySet::OnGetAll(response);
+
+    if (on_got_all_) {
+      std::move(on_got_all_).Run();
+    }
+  }
+
+ private:
+  base::OnceCallback<void()> on_got_all_;
+};
+
+GeoClueLocationProvider::GeoClueLocationProvider() {
+  DETACH_FROM_SEQUENCE(sequence_checker_);
+
+  dbus::Bus::Options options;
+  options.bus_type = dbus::Bus::SYSTEM;
+  options.connection_type = dbus::Bus::PRIVATE;
+  options.dbus_task_runner = dbus_thread_linux::GetTaskRunner();
+
+  bus_ = base::MakeRefCounted<dbus::Bus>(options);
+}
+
+GeoClueLocationProvider::~GeoClueLocationProvider() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
```

Comment:
it seems this class is only called on one sequence and `DCHECK_CALLED_ON_VALID_SEQUENCE` is typically only used when methods are not all called on the same sequence